tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/KnowledgeGuardian.scala

```diff
@@ -73,6 +78,52 @@ private[emitter] final class KnowledgeGuardian {
     invalidateAll
   }
 
+  private def computeHasInlinedInit(linkingUnit: LinkingUnit): Set[String] = {
+    /* Those classes are instantiated in scalajsenv.js. Since they have
+     * multiple constructors and/or are not final, scalajsenv.js is written
+     * with the assumption that they will not have an inlined init. We
+     * therefore blacklist them here so that this is always true.
+     */
+    val blackList = Set(
+        "jl_ClassCastException",
+        "jl_ArrayIndexOutOfBoundsException",
+        "sjsr_UndefinedBehaviorError",
+        "js_CloneNotSupportedException"
+    )
+
+    val scalaClassDefs = linkingUnit.classDefs.filter(_.kind.isClass)
+    val instantiatedClasses = scalaClassDefs.filter(_.hasInstances)
+    val classesWithInstantiatedSubclasses =
+      instantiatedClasses.flatMap(_.superClass).map(_.name).toSet
+
+    val cantHaveInlinedInit = classesWithInstantiatedSubclasses ++ blackList
+
+    def enableInlinedInitFor(classDef: LinkedClass): Boolean = {
+      /* We can enable inlined init if all of the following apply:
+       * - The class is not blacklisted
+       * - It does not have any instantiated subclass
+       * - It has exactly one (regular) constructor
+       * - It does not have an exported constructor
```

Comment:
I like this list. I'd consider not constructing the intermediate set `cantHaveInlinedInit` and just add two set lookups.